src/highdicom/pr/sop.py

```diff
@@ -0,0 +1,1459 @@
+"""Module for SOP Classes of Presentation State (PR) IODs."""
+from collections import defaultdict
+import datetime
+from typing import Sequence, Optional, Tuple, Union
+
+from PIL.ImageCms import ImageCmsProfile, createProfile
+
+from pydicom import Dataset
+from pydicom.sr.coding import Code
+from pydicom.uid import ExplicitVRLittleEndian
+from pydicom._storage_sopclass_uids import (
+    GrayscaleSoftcopyPresentationStateStorage,
+    PseudoColorSoftcopyPresentationStateStorage,
+    ColorSoftcopyPresentationStateStorage
+)
+from pydicom.valuerep import DA, PersonName, TM, format_number_as_ds
+
+import numpy as np
+
+from highdicom.base import SOPClass
+from highdicom.content import (
+    ContentCreatorIdentificationCodeSequence,
+    LUT,
+    ModalityLUT,
+    PaletteColorLookupTable,
+    ReferencedImageSequence,
+)
+from highdicom.pr.content import (
+    GraphicLayer,
+    GraphicGroup,
+    GraphicAnnotation,
+    SoftcopyVOILUT,
+)
+from highdicom.pr.enum import PresentationLUTShapeValues
+from highdicom.enum import RescaleTypeValues
+from highdicom.sr.coding import CodedConcept
+from highdicom.uid import UID
+from highdicom.valuerep import (
+    check_person_name,
+    _check_code_string,
+    _check_long_string
+)
+
+
+class _SoftcopyPresentationState(SOPClass):
+
+    """An abstract base class for various Presentation State objects."""
+
+    def __init__(
+        self,
+        referenced_images: Sequence[Dataset],
+        series_instance_uid: str,
+        series_number: int,
+        sop_instance_uid: str,
+        sop_class_uid: str,
+        instance_number: int,
+        manufacturer: str,
+        manufacturer_model_name: str,
+        software_versions: Union[str, Tuple[str]],
+        device_serial_number: str,
+        content_label: str,
+        content_description: Optional[str] = None,
+        graphic_annotations: Optional[Sequence[GraphicAnnotation]] = None,
+        graphic_layers: Optional[Sequence[GraphicLayer]] = None,
+        graphic_groups: Optional[Sequence[GraphicGroup]] = None,
+        concept_name_code: Union[Code, CodedConcept, None] = None,
+        institution_name: Optional[str] = None,
+        institutional_department_name: Optional[str] = None,
+        content_creator_name: Optional[Union[str, PersonName]] = None,
+        content_creator_identification: Optional[
+            ContentCreatorIdentificationCodeSequence
+        ] = None,
+        rescale_intercept: Union[int, float, None] = None,
+        rescale_slope: Union[int, float, None] = None,
+        rescale_type: Union[RescaleTypeValues, str, None] = None,
+        modality_lut: Optional[ModalityLUT] = None,
+        copy_modality_lut: bool = False,
```

Comment:
The standard says that the Modality LUT module is required if "a Modality LUT is to be applied to referenced image(s)". Does that mean that if the presentation state does not specify a Modality LUT, none should be applied even if the source image specifies one?

Reply:
I find these parameters confusing. Could we instead add an alternative constructor method to the `ModalityLUT` and `SoftcopyVOILUT` classes to facilitate the creation of objects from an existing image?

We have taken a similar approach with [highdicom.sr.SourceImageForSegmentation.from_source_image()](https://github.com/herrmannlab/highdicom/blob/b9a6a3ebf4a34b69a627b2785565623627fdf62e/src/highdicom/sr/content.py#L443).

Reply:
In addition, we would need a similar logic for ICC Profile of color images. To avoid a proliferation of `copy_*` attributes, I would let the caller deal with that beforehand

Reply:
We have seen that at least some viewers (eUnity) do not apply any Modality or VOI transforms if they are present in the image but then not in the presentation state.  This leads to basically unviewable images. I think the standard is ambiguous on this matter, although maybe the strictest reading would lead to that conclusion (rather silly in my opinion). Hence I added these options to conveniently copy them from the original images. The alternative constructor method will not work because sometimes multiple items are needed to match the effect of the original transforms...

Reply:
We could create a utility function (rather than a constructor) to create the correct objects given a list of existing image datasets. But this just seems less convenient and I can't see the benefit. Why do you find the `copy_*_lut` options confusing? Can we maybe just clarify them somehow?

Reply:
As pointed out by @dclunie, [Part 04 N.2 Pixel Transformation Sequence](https://dicom.nema.org/medical/dicom/current/output/chtml/part04/sect_N.2.html#para_2be00680-9512-4f36-8ca9-d398f5fffe1b) states the following:

> The transformations defined in the Softcopy Presentation State Storage SOP Classes replace those that may be defined in the Referenced Image SOP Instance. If a particular transformation is absent in the Softcopy Presentation State Storage SOP Class, then it shall be assumed to be an identity transformation, and any equivalent transformation, if present, in the Referenced Image SOP Instance shall NOT be used instead.

Reply:
It appears that the presentation state completes overrides any "styling" information contained in the source image. If no transformation is specified by the presentation state, no transformation should be applied.

Reply:
> We could create a utility function (rather than a constructor) to create the correct objects given a list of existing image datasets. But this just seems less convenient and I can't see the benefit. Why do you find the copy_*_lut options confusing? Can we maybe just clarify them somehow?

On the one hand, it's convenient for the caller to just set `copy_modality_lut` to `True` and let the constructor handle everything. There is also a slight risk when the information needed for `ModalityLUT` is obtained elsewhere because there is a chance that the information is copied from a different image instead of the one referenced in the presentation state object. However, one could say the same about most of the abstractions in the `sr` package.

On the other hand, I feel that it would be cleaner if the caller would have to explicitly specify these transformations without the presentation state class making any assumptions. The constructor can then focus on verifying that all required information is provided and raise an exception if something is missing or invalid.

The logic of the pixel transformation pipeline using Modality LUT, VOI LUT, Palette Color Lookup Table, and ICC Profile is quite intricate and appears to differ slightly between image modalities (slightly different modules, macros, etc.), and I am afraid that the presentation state class will need to take on more and more complexity for handling all edge cases correctly. Also, if the information in the image has some compliance issues that would need to be fixed before copying, one would need additional parameters such as `fix_modality_lut_before_copying`. That's not really the purpose of the constructor in my opinion and results in low cohesion. Therefore, I would rather prefer an alternative constructor method of `ModalityLUT` (e.g., `ModalityLUT.from_image()`), which could be further parametrized if needed.

Reply:
Right, but that is almost never what someone actually wants to do I would argue, hence adding these convenience options.

When we were playing around with CTs and not including any Modality LUT or VOI LUT resulted in very strange presentation of the images to which the presentation state was applied, they were all saturated and basically unreadable, as you'd expect for a "raw" CT image. Hence I think these options are basically essential for the library to be usable.

Reply:
(posted my last reply before i had reloaded your previous message so I was replying to the previous message, still thinking about your last message)

Reply:
> When we were playing around with CTs, not including any Modality LUT or VOI LUT resulted in very strange presentation of the images to which the presentation state was applied in viewers: they were all saturated and basically unreadable, as you'd expect for a "raw" CT image. Hence I think these options are basically essential for the library to be usable. By just specifying one option in the constructor, the user can very easily create usable presentation states. Dealing with this themselves would be a nightmare.

Fully agreed! The question is only where this should be handled and whether that handling may need additional parameters.

I have painful memories of `highdicom.sr.ContentSequence.from_sequence()`, where we initially thought this was straight forward, and then needed `is_root` and `is_sr` to handle various corner cases.

Reply:
See 990bf06b176612fc92bbf580d7e681a486a32bc2

Reply:
See 1df7f01404b14e48bd97b2cf56d249bdf02c3daa for the handling of ICC profiles